code/modules/crafting/craft.dm

```diff
@@ -151,152 +151,125 @@
-							SD = new S.type()
-							Deletion += SD
-						S.use(amt)
-						SD = locate(S.type) in Deletion
-						SD.amount += amt
-						continue main_loop
-					else
-						amt -= S.amount
-						if(!locate(S.type) in Deletion)
-							Deletion += S
-						else
-							data = S.amount
-							S = locate(S.type) in Deletion
-							S.add(data)
-						surroundings -= S
-			else
-				var/atom/movable/I
-				while(amt > 0)
-					I = locate(A) in surroundings
-					Deletion += I
-					surroundings -= I
-					amt--
-	var/list/partlist = list(R.parts.len)
-	for(var/M in R.parts)
-		partlist[M] = R.parts[M]
-	for(var/A in R.parts)
-		if(istype(A, /datum/reagent))
-			var/datum/reagent/RG = locate(A) in Deletion
-			if(RG.volume > partlist[A])
-				RG.volume = partlist[A]
-			. += RG
-			Deletion -= RG
-			continue
-		else if(istype(A, /obj/item/stack))
-			var/obj/item/stack/ST = locate(A) in Deletion
-			if(ST.amount > partlist[A])
-				ST.amount = partlist[A]
-			. += ST
-			Deletion -= ST
-			continue
+	if(!do_after(user, recipe.time, target = user))
+		return "."
+	contents = get_surroundings(user)
+
+	if(!check_contents(recipe, contents))
+		return ", missing component."
+	if(!check_tools(user, recipe, contents))
+		return ", missing tool."
+	if(!check_pathtools(user, recipe, contents))
+		return ", missing tool."
+	
+	var/list/parts = del_reqs(recipe, user)
+	if(!parts)
+		return ", missing component."
+					
+	var/atom/movable/craft_result = new recipe.result (get_turf(user.loc))
+	craft_result.CheckParts(parts, recipe)
+	if(isitem(craft_result))
+		user.put_in_hands(craft_result)
+
+	if(send_feedback)
+		SSblackbox.record_feedback("tally", "object_crafted", 1, craft_result.type)
+	return 0
+			
+
+/datum/personal_crafting/proc/del_reqs(datum/crafting_recipe/recipe, mob/user)
+	var/list/surroundings = get_environment(user)
+	var/list/parts_list_for_return = list()
+	var/list/things_for_deletion = list()
+
+	for(var/thing in recipe.reqs)
+		var/needed_amount = recipe.reqs[thing]
+		if(ispath(thing, /datum/reagent))
+			var/datum/reagent/part_reagent = locate(thing) in parts_list_for_return
+			if(!part_reagent)
+				part_reagent = new thing()
+				parts_list_for_return += part_reagent
+
+			for(var/obj/item/reagent_containers/container in surroundings)
+				var/datum/reagent/contained_reagent = container.reagents.get_reagent(thing)
+				if(!contained_reagent)
+					continue
+				
+				if(contained_reagent.volume > needed_amount)
+					things_for_deletion[container] = list(contained_reagent, needed_amount)
+					part_reagent.volume = needed_amount
+					part_reagent.data = contained_reagent.data
+					needed_amount = 0
+					break
+				else
+					needed_amount -= contained_reagent.volume
+					things_for_deletion[container] = list(contained_reagent, contained_reagent.volume)
+					part_reagent.volume += contained_reagent.volume
+					part_reagent.data += contained_reagent.data
+	
+			if(needed_amount > 0)
+				return null
```

Comment:
If we reach this point, something has gone wrong (because we must have CHECKED beforehad that we have enough stuffs!), and problems probably arose (we already deleted reagents/stacks, while iterating through previous reqs).
This is a kind of "this should never happen" situation in which i'd think a runtime would be appropriate.